### PR TITLE
WIP: Draw backgrounds separately from foregrounds

### PR DESCRIPTION
**Pull request by DHowett**
This implements #6193. It's a WIP.

## Summary of the Pull Request

---

Alright, I've got this down to a single function that takes each dirty region and produces...

* A list of damaged backgrounds to repaint
* A list of colored cluster lists to repaint

I don't like the number of vectors here, but... we're saving money by invalidating narrowly whenever possible.

```
Dirty Region [1..n]
- Origin (SCREEN CHAR COORDS)
- At line wrap point
- Region intersected right half of DBCS char (left trim)
- Background Run [1..n]
  - Attribute
  - Rect Covered (SCREEN CHAR COORDS)
- Colored Foreground Run [1..n]
  - Attribute
  - Columns
  - Cluster [1..n]
```

Right now, because of some minor issues (#2661), backgrounds are attributes _and_ actual final colors. It's not pleasant!

To fix:

* [ ] reverse is occasionally (!) broken
* [x] render engines need flags to say "plz no background splitting" (vt engine needs to update drawing brushes per text run, foreground _and_ background)
* [ ] need to re-enable space optimization (if the cell is non-printing (and has no underline, etc.), no point in forwarding it to the cluster renderer!)

---

 
## PR Checklist
* [ ] Closes #xxx
* [ ] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/Terminal) and sign the CLA
* [ ] Tests added/passed
* [ ] Requires documentation to be updated
* [ ] I've discussed this with core contributors already. If not checked, I'm ready to accept this work might be rejected in favor of a different grand plan. Issue number where discussion took place: #xxx

### CODE REVIEW

File: src/renderer/base/renderer.cpp [hunk 1]
```diff
@@ -632,46 +819,51 @@ void Renderer::_PaintBufferOutput(_In_ IRenderEngine* const pEngine)
...
                                         TextBufferCellIterator it,
                                         const COORD target,
                                         const bool lineWrapped)
 {
-    auto globalInvert{ _pData->IsScreenReversed() };
+    (void)lineWrapped;
+    (void)target;
+    (void)it;
+    (void)pEngine;
+#if 0
```
Comment (DHowett) [inline src/renderer/base/renderer.cpp:865]: these, and their guarded regions, wil be deleted

File: src/buffer/out/TextAttribute.hpp [hunk 2]
```diff
@@ -154,22 +154,29 @@ class TextAttribute final
...
         // inverted is whether there's a global invert; Reverse is a local one.
         // global ^ local == true : the background attribute is actually the visible foreground, so we care about the foregrounds being identical
         // global ^ local == false: the foreground attribute is the visible foreground, so we care about the backgrounds being identical
         const auto checkForeground = (inverted != IsReverseVideo());
-        return !IsAnyGridLineEnabled() && // grid lines have a visual representation
+        return ((checkForeground && _foreground == other._foreground) ||
+                (!checkForeground && _background == other._background));
+    }
+
+    // This returns whether this attribute, if printed directly next to another attribute, for the space
```
Comment (DHowett) [inline src/buffer/out/TextAttribute.hpp:168]: wrong. need to resurrect this tho

**Comment by DHowett** [comment]
This will be cut deep by the 2661 fix.,

File: src/renderer/base/renderer.cpp [hunk 3]
```diff
@@ -104,6 +104,53 @@ Renderer::~Renderer()
     return S_OK;
 }
 
+template<>
+struct fmt::formatter<TextColor>
```
Comment (DHowett) [inline src/renderer/base/renderer.cpp:108]: don't need these to be in a generic place, but we could move them into their appropriate headers later

**Comment by DHowett** [comment]
Merge in work from the #803 bugfix -- *also*, maybe we should generalize BackgroundRun to RenderableAttributeRun! This would let us katamari together runs line drawing characters and break them _only when they change_.

We want to consider how this interacts with underline (part of foreground).

**Comment by miniksa** [comment]
reopening and assigning to myself, I might need this for perf.

**Comment by DHowett** [comment]
Don't need to keep open, will post branch name in issue

**Pull request closed without merging**